### PR TITLE
Allow specifying CustomJS using a Python function

### PR DESCRIPTION
**Pull request by almarklein**
Issues: implements part 1 of #2788

This uses ``flexx.pyscript`` to transpile a given Python function to JS. Except for docs and tests this is all there is to it.

However, we may want to wait until PyScript has evolves a bit further, especially Python 2.7 support.

### CODE REVIEW

**Comment by damianavila** [comment]
> However, we may want to wait until PyScript has evolves a bit further, especially Python 2.7 support.

This is **super** interesting... but do they really want to support 2.7??

**Comment by mattpap** [comment]
> (...) but do they really want to support 2.7?

Yes.

**Comment by almarklein** [comment]
>  but do they really want to support 2.7??

What do you mean? I'm the author :)  Principally, I would prefer not to bother with Python 2.x, but as a Bokeh dev I would be hesitant to use it if it only ran on 3.x.

**Comment by damianavila** [comment]
>What do you mean? I'm the author :) 

JAJA... I did not have time to check flexx in detail... I completely missed you are the author, I just read a little bit of the docs and never found your name...

>I would prefer not to bother with Python 2.x, but as a Bokeh dev I would be hesitant to use it if it only ran on 3.x.

I understand... probably I am a little bit biased, but since some time ago, for my own projects I don't bother if they are not py2 compatible... because py2 is the past and I want to move forward... so I would completely understand if people is reluctant to make their work py2 compat...

**Comment by fpliger** [comment]
First, this is super exciting!

> I would prefer not to bother with Python 2.x, but as a Bokeh dev I would be hesitant to use it if it only ran on 3.x.

Agreed. FWIW supporting just Python 3 as a first iteration of the feature and extending support to Python 2.7 later is an acceptable path. The risk is that "later" can tend to never... but still having the feature only for Python 3 is better then not having it at all.

**Comment by almarklein** [comment]
While we're on the subject, does Bokeh support anything other than cPython? E.g. PyScript also does not yet work with Pypy.

**Comment by bryevdv** [comment]
We tested it with pypy for fun a long time ago and it worked for the casual uses we tried by hand. It's not something we currently maintain under test so I'm not sure if some or all the functionality still works.

**Comment by bryevdv** [comment]
FYI Travis and several others seem favorable to releasing this as a "python 3 only" feature. That wasn't my first inclination, but I can see the value in it.

**Comment by almarklein** [comment]
Cool, then Flexx can stay py3k only. In that case the only feature that I think is missing from PyScript is Pythonic operators (e.g. so ``'3'*3`` becomes "333" and not 9).

**Comment by datnamer** [comment]
Really excited about this, thanks!

**Comment by bryevdv** [comment]
Well, I see that I forgot to add "for now" which would certainly still be my preference, if at all possible.

**Comment by bryevdv** [comment]
@almarklein can you give estimates for: 

* resolving the conflicts in the current branch
* adding 2.7 support

I think it can go in 0.11 as "py 3 only" but there will be demand for 2.7 support for sure.

**Comment by almarklein** [comment]
I plan to get on this as soon as I finish with the webgl line. The conflict itself should be easily resolved by adding flexx as a dependency to Travis. However, there's a few improvements that I'd like to make to PyScript (getting CI up being one of them) to feel confident about making this available in Bokeh. So maybe end next week?

I'm actually increasingly liking the idea to keep Flexx Py3k only :)  Having said that I am interested in making PyScript work on Pypy and maybe ... PyScript, so perhaps the step to get it working in Python 2.7 will be almost free at some point. Let's start with py3k and discuss this later on.

File: examples/plotting/file/slider.py [hunk 1]
```diff
@@ -14,19 +14,19 @@
...
+    data = source.get('data')
+    A = amp.get('value')
+    k = freq.get('value')
+    phi = phase.get('value')
+    B = offset.get('value')
+    x, y = data['x'], data['y']
+    for i in range(len(x)):
+        y[i] = B + A*Math.sin(k*x[i]+phi)
+    source.trigger('change') 
+
```
Comment (havocp) [inline examples/plotting/file/slider.py]: Is there any way we could make this API match the Python API for HasProps/PlotObject? It looks like the main thing would be to change `source.data` in Python to `source.get('data')` in JS.

Comment (almarklein) [inline examples/plotting/file/slider.py]: That's a good point. In Flexx I've made sure from the start that the API's are equal. In Bokeh we could perhaps add a transformation step to make it equal. Or ... change the JS API to better reflect the Python API. We'll have to look into that.

**Comment by almarklein** [comment]
This diverged. new PR in a bit

**Comment by almarklein** [comment]
#3121

**Pull request closed without merging**